Source/Scene/ClippingPlanesCollection.js

```diff
@@ -0,0 +1,245 @@
+         * @default Matrix4.IDENTITY
+         */
+        this.modelMatrix = defaultValue(options.modelMatrix, Matrix4.clone(Matrix4.IDENTITY));
+
+        /**
+         * The color applied to highlight the edge along which an object is clipped.
+         *
+         * @type {Color}
+         * @default Color.WHITE
+         */
+        this.edgeColor = defaultValue(options.edgeColor, Color.clone(Color.WHITE));
+
+        /**
+         * The width, in pixels, of the highlight applied to the edge along which an object is clipped.
+         *
+         * @type {Number}
+         * @default 0.0
+         */
+        this.edgeWidth = defaultValue(options.edgeWidth, 0.0);
+
+        this._testIntersection = undefined;
+        this._combineClippingRegions = undefined;
+        this.combineClippingRegions = defaultValue(options.combineClippingRegions, true);
+    }
+
+    defineProperties(ClippingPlanesCollection.prototype, {
+        /**
+         * If true, the region to be clipped must be included in all planes in this collection.
+         * Otherwise, a region will be clipped if included in any plane in the collection.
+         *
+         * @memberof ClippingPlanesCollection.prototype
+         * @type {Boolean}
+         * @default true
+         */
+        combineClippingRegions : {
+            get : function() {
+                return this._combineClippingRegions;
+            },
+            set : function(value) {
+                if (this._combineClippingRegions !== value) {
+                    this._combineClippingRegions = value;
+                    this._testIntersection = getTestIntersectionFunction(value);
+                }
+            }
+        }
+    });
+
+    function getTestIntersectionFunction(combineClippingRegions) {
+        if (combineClippingRegions) {
+            return function(value) {
+                return (value === Intersect.INSIDE);
+            };
+        }
+
+        return function(value) {
+            return (value === Intersect.OUTSIDE);
+        };
+    }
+
+    var scratchPlane = new Plane(Cartesian3.UNIT_X, 0.0);
+    var scratchMatrix = new Matrix4();
+    /**
+     * Applies the transformations to each plane and packs it into an array.
+     * @private
+     *
+     * @param viewMatrix The 4x4 matrix to transform the plane into eyespace.
+     * @param [array] The array into which the planes will be packed.
+     * @returns {Cartesian4[]} The array of packed planes.
+     */
+    ClippingPlanesCollection.prototype.transformAndPackPlanes = function(viewMatrix, array) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.typeOf.object('viewMatrix', viewMatrix);
+        //>>includeEnd('debug');
+
+        var planes = this.planes;
+        var length = planes.length;
+
+        var i;
+        if (!defined(array) || array.length !== length) {
+            array = new Array(length);
+
+            for (i = 0; i < length; ++i) {
+                array[i] = new Cartesian4();
```

Comment:
We can be smarter about this.  You can assume that the result array is already filled with `Cartesian4` right?  So we only need to create new ones for the range `i = oldArrayLength.length; i < length; i++`
If the array passed in is too large, we can simply do `array.length = length`

Reply:
Why are we using a `Cartesian4` anyway?  Wouldn't it be better just to use a flat array where every 4 indices is equal to 1 plane?

Reply:
> You can assume that the result array is already filled with Cartesian4 right?

We changed to this because it's a public method, and the result parameter is optional like our other methods that take result paramters. We can require it though if that's ok.

Reply:
It's alright if the result parameter is optional. But if a result parameter is supplied, we should be able to assume [0..length] is already filled with `Cartesian4` so we don't need to make a new one for those indices